api/envoy/extensions/filters/http/cache/v3/cache.proto

```diff
@@ -49,9 +50,13 @@ message CacheConfig {
     repeated config.route.v3.QueryParameterMatcher query_parameters_excluded = 4;
   }
 
-  // Config specific to the cache storage implementation.
+  // Config specific to the cache storage implementation. Required unless ``disabled``
+  // is true.
   // [#extension-category: envoy.http.cache]
-  google.protobuf.Any typed_config = 1 [(validate.rules).any = {required: true}];
+  google.protobuf.Any typed_config = 1;
+
+  // When true, the cache filter is a no-op filter.
+  google.protobuf.BoolValue disabled = 5;
```

Comment:
Instead of adding this field, why not just say that if the `typed_config` field is unset, the filter is disabled?  That approach seems more consistent with how other filters work.

I realize that a control plane sending a config without that field set would break older versions of Envoy due to the PGV annotation that you're changing here.  However, those versions of Envoy also wouldn't know about this new field, so you wouldn't really be able to use the ability to disable the filter with those versions of Envoy anyway.

Reply:
That was one of the options considered early on, but there are several advantages of this:

1. Disable by omission is a bit non-explicit - it would be very unclear looking at config that `cache {}` means don't cache. `cache {disabled:true}` is pretty clear what's going to happen.

2. That model wouldn't be forward-compatible with the idea of inherited per-route config - there would be no way to signify "I am intentionally turning off the cache for this route" because unset means inherit. (i.e. listener `typed_config=file_system_cache`, route `typed_config=unset` would result in inheriting the listener's cache config.)

3. Doing it this way also makes possible the desirable inverse of that behavior (assuming future implementation of inheritance-based config); one can do e.g. at the listener level `typed_config=file_system_cache{...}, disabled=true`, and then per-route `disabled=false` to enable and inherit the top-level cache config per route, without repeating that potentially large config.

It was also considered to do "child configs 100% override parent configs", which would make configuring "off by omission" unambiguous again, but apparently other cache users specifically want to be able to do inherited config, so the PR that was going to do it as full config overrides was shot down.

Reply:
(I also previously suggested `oneof {typed_config / disabled}`, but wrapping an existing field into a oneof isn't backwards-compatible in golang. That model also wouldn't play well with point 3 above, anyway.)

Reply:
> 1. Disable by omission is a bit non-explicit - it would be very unclear looking at config that `cache {}` means don't cache. `cache {disabled:true}` is pretty clear what's going to happen.

I don't actually think that's unclear, and it seems consistent with what other filters do.  For example, in the RBAC filter config, if the [`rules` field](https://github.com/envoyproxy/envoy/blob/284be839e4585507d38af6f067efbc7fb14512cd/api/envoy/extensions/filters/http/rbac/v3/rbac.proto#L34) is unset, that disables the filter.

> 2. That model wouldn't be forward-compatible with the idea of inherited per-route config - there would be no way to signify "I am intentionally turning off the cache for this route" because unset means inherit. (i.e. listener `typed_config=file_system_cache`, route `typed_config=unset` would result in inheriting the listener's cache config.)
> 3. Doing it this way also makes possible the desirable inverse of that behavior (assuming future implementation of inheritance-based config); one can do e.g. at the listener level `typed_config=file_system_cache{...}, disabled=true`, and then per-route `disabled=false` to enable and inherit the top-level cache config per route, without repeating that potentially large config.
> It was also considered to do "child configs 100% override parent configs", which would make configuring "off by omission" unambiguous again, but apparently other cache users specifically want to be able to do inherited config, so the PR that was going to do it as full config overrides was shot down.

I actually think "child configs 100% override parent configs" is probably a better model.  I agree that it can make some cases more verbose due to needing to copy config into multiple routes.  However, I think the semantics of figuring out how to merge different fields between the top-level config and the override config can wind up making it much harder for humans to understand the config.  To say that another way, I think it would be better for the `cache` field being unset to mean disable than for it to mean inherit.

@envoyproxy/api-shepherds, do we actually have any other filters that merge configs on a field-by-field basis like this?  If there is precedent for this, I'd be okay with it, but I'd prefer not to introduce this approach if we're not already doing it in existing filters.

Reply:
> @envoyproxy/api-shepherds, do we actually have any other filters that merge configs on a field-by-field basis like this? If there is precedent for this, I'd be okay with it, but I'd prefer not to introduce this approach if we're not already doing it in existing filters.

I'm aware of both 100% override (e.g., [Buffer](https://github.com/envoyproxy/envoy/blob/9adaaf6d40370ca4998829d343cf4dcdaa58aa36/api/envoy/config/filter/http/buffer/v2/buffer.proto#L39), [StatefulSession](https://github.com/envoyproxy/envoy/blob/9adaaf6d40370ca4998829d343cf4dcdaa58aa36/api/envoy/extensions/filters/http/stateful_session/v3/stateful_session.proto#L37)) and field-by-field override (e.g., [ext_proc](https://github.com/envoyproxy/envoy/blob/9adaaf6d40370ca4998829d343cf4dcdaa58aa36/api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto#L181), [ext_auth](https://github.com/envoyproxy/envoy/blob/9adaaf6d40370ca4998829d343cf4dcdaa58aa36/api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto#L336)).

From the point of view of a reader/maintainer of a config, I think a 100% override makes life easier (it is fully described in one place, and no need to worry what happens if some other person updates the config at the higher level).
From "performance" point of view (size of messages sent over the wire and processed) it is a downside, especially for large filter configurations.

+1 on clarifying the comment. Adding the use-case where this is useful (common configuration that can be overridden by per-route configs) to the comment will help here.

Reply:
I would like to add the explanation to the comment, but I think that would be *more* confusing right now because there is no per-route config support in this filter.

I already had a PR enabling per-route-config shot down in part because it *didn't* support field-by-field overrides, which apparently the Google folks consider to be an important feature for this filter. They specifically wanted to be able to inherit an impl config, and override only some of the not-yet-implemented fields.

I agree with 100% overrides being easier to understand, and they're also right now a thousand times easier to implement; the issue raised by Todd is very large configs. I'm also on board with inheritance-based configs if that's how we want to do it. I also agree that it's a problem that right now we have a mix of implementations, and there's no way to unify on one method at this point.

For me, the bigger problem is I can't get [that PR](https://github.com/envoyproxy/envoy/pull/25825) accepted because some reviewers want support for inherited config, and I can't get this PR accepted because other reviewers think it's better not to even potentially support inherited config!

(For prior discussion of supporting inherited config, see also #25903 and [related maintainer-only slack thread](https://envoyproxy.slack.com/archives/G79D81J7Q/p1677788321804009).)

Reply:
I didn't actually realize that we already had some filters that do per-field override.  If that pattern already exists, then I don't think there's a good reason not to go with that approach here.  My concern was not wanting to introduce this as a new pattern if it didn't already exist, but that's obviously not the case here.

So I think this is fine.  Sorry for the noise!